### PR TITLE
Bugfix/sign/respond session request catch error

### PR DESCRIPTION
**Pull request by TalhaAli00**

### CODE REVIEW

File: protocol/sign/src/main/kotlin/com/walletconnect/sign/engine/use_case/calls/DisconnectSessionUseCase.kt [hunk 1]
```diff
@@ -18,11 +18,11 @@ import kotlinx.coroutines.supervisorScope
 internal class DisconnectSessionUseCase(
     private val jsonRpcInteractor: JsonRpcInteractorInterface,
     private val sessionStorageRepository: SessionStorageRepository,
-    private val logger: Logger
+    private val logger: Logger,
 ) : DisconnectSessionUseCaseInterface {
     override suspend fun disconnect(topic: String, onSuccess: () -> Unit, onFailure: (Throwable) -> Unit) = supervisorScope {
         if (!sessionStorageRepository.isSessionValid(Topic(topic))) {
-            throw CannotFindSequenceForTopic("$NO_SEQUENCE_FOR_TOPIC_MESSAGE$topic")
+            return@supervisorScope onFailure(CannotFindSequenceForTopic("$NO_SEQUENCE_FOR_TOPIC_MESSAGE$topic"))
```
Comment (jakubuid) [inline protocol/sign/src/main/kotlin/com/walletconnect/sign/engine/use_case/calls/DisconnectSessionUseCase.kt:25]: What's the reason of this change?

File: protocol/sign/src/main/kotlin/com/walletconnect/sign/engine/use_case/calls/ExtendSessionUseCase.kt [hunk 2]
```diff
@@ -19,23 +19,23 @@ import com.walletconnect.sign.common.model.vo.clientsync.session.params.SignPara
...
         }
 
         val session = sessionStorageRepository.getSessionWithoutMetadataByTopic(Topic(topic))
         if (!session.isSelfController) {
-            throw UnauthorizedPeerException(UNAUTHORIZED_EXTEND_MESSAGE)
+            return@supervisorScope onFailure(UnauthorizedPeerException(UNAUTHORIZED_EXTEND_MESSAGE))
         }
         if (!session.isAcknowledged) {
-            throw NotSettledSessionException("$SESSION_IS_NOT_ACKNOWLEDGED_MESSAGE$topic")
+            return@supervisorScope onFailure(NotSettledSessionException("$SESSION_IS_NOT_ACKNOWLEDGED_MESSAGE$topic"))
```
Comment (jakubuid) [inline protocol/sign/src/main/kotlin/com/walletconnect/sign/engine/use_case/calls/ExtendSessionUseCase.kt:38]: The same question

**Review by jakubuid — COMMENTED** [review]
What's the purpose of this change? Exceptions thrown in use cases should be already caught in SignProtocol class and exposed in onError callback to devs

**Review by kacperoak — APPROVED** [review]

**Comment by TalhaAli00** [comment]
> What's the purpose of this change? Exceptions thrown in use cases should be already caught in SignProtocol class and exposed in onError callback to devs

This is to fix the issue Trust is seeing in their app. They are currently seeing around 2k crashes being reported and the stacktrace was shared in #1170. 

The protocol class doesn't capture the exceptions that are thrown and it causing crashes at runtime. I'm not 100% but my guess is that because we're launching our usecases in a supervisor scope, that is why the exceptions aren't being propagated back to the protocol class.

**Comment by jakubuid** [comment]
> > What's the purpose of this change? Exceptions thrown in use cases should be already caught in SignProtocol class and exposed in onError callback to devs
> 
> This is to fix the issue Trust is seeing in their app. They are currently seeing around 2k crashes being reported and the stacktrace was shared in #1170.
> 
> The protocol class doesn't capture the exceptions that are thrown and it causing crashes at runtime. I'm not 100% but my guess is that because we're launching our usecases in a supervisor scope, that is why the exceptions aren't being propagated back to the protocol class.

Hmm that's interesting. Have you been able to reproduce this behavior? In my testing, I have not been able to.

Still, to me, it looks like a way of mitigating the issue not solving the root cause. Do we know why Trust is seeing `CannotFindSequenceForTopic` exception? Do we know how to repro?

I did my investigation: https://github.com/WalletConnect/WalletConnectKotlinV2/pull/1175
Based on:
https://github.com/WalletConnect/WalletConnectKotlinV2/issues/1105
https://github.com/WalletConnect/WalletConnectKotlinV2/issues/1130

I believe these might be connected

**Comment by muratogat** [comment]
Perhaps I'm a bit a bit out of place here, but I just wanted to confirm that :
- IMO #1105 is the exact reason Trust or any other wallets are experiencing `CannotFindSequenceForTopic `
- I also don't see any of the exceptions being propageted to onError and lead to a crash instead. Even wrapping the block for approving-rejecting in try-catch doesn't work in that case. That's why I manually check if the topic is present before attempting any actions, as described [here](https://github.com/WalletConnect/WalletConnectKotlinV2/issues/1130#issuecomment-1742563971).

File: protocol/sign/src/main/kotlin/com/walletconnect/sign/engine/use_case/calls/DisconnectSessionUseCase.kt (hunk 1 above)
Comment (TalhaAli00) [reply protocol/sign/src/main/kotlin/com/walletconnect/sign/engine/use_case/calls/DisconnectSessionUseCase.kt:25]: this will propagate the error as before it was just being thrown

File: protocol/sign/src/main/kotlin/com/walletconnect/sign/engine/use_case/calls/ExtendSessionUseCase.kt (hunk 2 above)
Comment (TalhaAli00) [reply protocol/sign/src/main/kotlin/com/walletconnect/sign/engine/use_case/calls/ExtendSessionUseCase.kt:38]: same as above

File: protocol/sign/src/test/kotlin/com/walletconnect/sign/engine/use_case/calls/DisconnectSessionUseCaseTest.kt [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    private val disconnectSessionUseCase = DisconnectSessionUseCase(jsonRpcInteractor, sessionStorageRepository, logger)
+
+    @Before
+    fun setUp() {
+        every { logger.error(any() as String) } answers { }
+        every { logger.error(any() as Exception) } answers { }
+    }
+
+    @After
+    fun tearDown() {
```
[the remark this answers is not in the document]
Comment (TalhaAli00) [reply protocol/sign/src/test/kotlin/com/walletconnect/sign/engine/use_case/calls/DisconnectSessionUseCaseTest.kt]: removed

**Review by jakubuid — APPROVED** [review]
